### PR TITLE
feat: add initial version of Domain Wallet Authentication CAIP

### PR DESCRIPTION
**Pull request by FedericoAmura**
This CAIP is an identity standard. It proposes a method to associate crypto-domains and other account names with authorization providers in order to provide a portable web3 experience.

### CODE REVIEW

**Comment by bumblefudge** [comment]
Hey there-- sorry for the slow response time, I have been AFK for 2 weeks straight.

This definitely sounds like a CAIP that people would use, and I am glad to get this merged as a Draft soon so that people can see it's in progress/prototype phase and chime in from ENSDAO and other NS communities and ecosystems. My only requests before it be merged as draft is that:

1. we clean up the formatting and terminology a tiny bit to maximize likelihood of good feedback coming from casual readers/reviewers, and
2. you make some decisions about chain-specificity, both within and beyond EVM. 

every other in-line comment and suggestion can be considered non-urgent and picked up later if you're in a hurry to merge! speaking of non-urgent, i'd also recommend you put each sentence on its own line, it helps a lot with wordsmithing through github, as diffs display as whole sentences :D

**Review by bumblefudge — COMMENTED** [review]
Quick initial review-- apologies if it's a little inconsistent or incomplete!

**Comment by arein** [comment]
Can y'all please add sequence diagrams so network calls & context switches become more clear?

**Comment by arein** [comment]
How does this handle multiple accounts? Would this require me to have multiple profiles? How does it handle if my account resolves to different addresses on different chains which is a more emerging scenario with the rise of contract accounts?

**Comment by davidlsneider** [comment]
> How does this handle multiple accounts? Would this require me to have multiple profiles? How does it handle if my account resolves to different addresses on different chains which is a more emerging scenario with the rise of contract accounts?

We discussed this extensively and decided that this is best handled by the authenticator, if a user has multiple accounts with a given authenticator. In other words, the spec calls for associating one authenticator with a given name and then that authenticator can have multiple accounts associated with it. For example, when metamask opens, a user can then pick a wallet from there.

**Comment by pedrouid** [comment]
@davidlsneider as you are aware I'm a big fan of this concept but I'm concerned about the authflows being too generic or not descriptive enough

Describing "extension", "mwp" and "wc" is not sufficient to establish a connection and it will essentially create more problems than it solves

I think this CAIP requires us to put together a consistent experience between these authflows such that it's "automatically" discovered rather than communicated as a response

My proposal is that we create a new CAIP that emulates the EIP-6963 experience but make it chain-agnostic and potentially leverages web standards such as window.postMessage for discoverability

Therefore you can overlap between wallets present in the web plus wallets associated with the identifier

Untitled - Frame 1 (1)

File: CAIPs/caip-x.md [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+discussions-to: https://github.com/ChainAgnostic/CAIPs/pull/275
+status: Draft
+type: Standards Track
+Standard created: 2024-04-29
+updated: 2024-04-29
+---
+
+# Abstract
+
+The Domain Wallet Authentication describes a method for linking a crypto domain with authentication methods or providers by adding an authenticator: JSON/URL field to the metadata of a crypto domain NFT. The standard also describes a method for application developers and web3 login modal providers to enable users to login with their domain name.
```
[the remark this answers is not in the document]
Comment (FedericoAmura) [reply CAIPs/caip-x.md]: Added!

File: CAIPs/caip-x.md [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+    {
+      "connection": "wc",
+      "URI": "http://www.authprovider.com/auth/{}"
+    }
+  ]
+}
+```
+
+Application will filter the authentication flows that are not supported and then try to execute in order. In the example:
+When used on a browser, it will first try to execute the authentication flow that claims for an injected extension like metamask. In this case we are not defining any wallet rdns so it will use `window.ethereum` instead of looking at a specific EIP-6963 compliant wallet. If it cannot find the extension, it will try with the next flow and open another browser tab pointing to the corresponding URI, automatically passing it the wallet connect URI as query param to establish a connection upon user confirmation
```
[the remark this answers is not in the document]
Comment (FedericoAmura) [reply CAIPs/caip-x.md]: Updated section with steps

Those are the same method as one is more generic than the other, but still the same.
The reasoning is that first you define where you can find the wallet (injected) and then specify which one it is from the ones that you can find there. If you don't specify exactly which one it is, then it should be the one at `window.ethereum` whichever that is

File: CAIPs/caip-x.md [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+The standard functions that must be implemented by any resolver are of the format:
+
+### Function to Resolve a Domain Name to an Address
+
+> Function: resolveName
+> Description: Resolves a given domain name to a blockchain address.
+> Input: name (String) - The domain name to be resolved.
+> Output: Address (String | null) - The blockchain address associated with the domain name, or null if no address is found.
+
+This function takes a string input representing the domain name and returns either the associated blockchain address as a string or null if the address cannot be found. This function must handle various types of blockchain addresses and be adaptable to different blockchain technologies.
```
[the remark this answers is not in the document]
Comment (FedericoAmura) [reply CAIPs/caip-x.md]: The name resolver decides which chain or database to use.
In case we are using ENS as name resolver, it would be Ethereum, likely mainnet. But it could be anything, such as:
- Solana with some other NFT naming system
- Social networks or services and their databases (imagine if X gives wallets to their users)

File: CAIPs/caip-x.md [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+The Authentication flow definition JSON must have the following schema
+```json
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "type": "object",
+  "properties": {
+    "address": {
+      "type": "string",
+      "description": "The Ethereum address requested by the user, prefixed by 0x."
```
[the remark this answers is not in the document]
Comment (FedericoAmura) [reply CAIPs/caip-x.md]: Added chain as defined in CAIP-2

File: CAIPs/caip-x.md [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+
+The actual standard used to store this "authenticator" text record will vary depending on the Crypto Domain NFT system used.  For example, on ENS, a ENSIP-5 record should be created, with a "key" of "authenticator".  You can lookup the authenticator on an ENS domain using the text(bytes32 node, string key) function where "node" is the ENS domain being queried and "key" is the string "authenticator".
+
+Example of a response to retrieve the "authenticator" text record:
+
+"https://www.authprovider.com/auth/{}"
+
+## Authentication flow definition
+
+The User can provide their authentication flow by providing a URL where Web3 applications can fetch it.
```
[the remark this answers is not in the document]
Comment (FedericoAmura) [reply CAIPs/caip-x.md]: I see your point, and agree. Most users won't ever deal with this

I believe the most common use case for this will be wallets that define their own address naming system and integrate with ENS, such as coinbase domains under `*.cb.id`. They provide naming, and could also provide `authenticator` text records, offline

The self hosting option or the feature to write it yourself in the `authenticator` text record still have to be there as those are the only options that would allow any user to not trust anybody and to have as much freedom as they want/need (maybe combining competing wallets)

Anyway we rephrased this section, hopefully it is more clear now what we wanted to express

**Comment by bumblefudge** [comment]
Hey all, sorry to be offline for so long, was on (actually still am on) a mini-sabbatical.

I share Pedro's concern that this flattens categories a bit, and I think the success or failure of the endeavor hinges on how tightly specified each authFlow "type" is.  Without diving too far into hypotheticals or doing any research, my first impulse is-- why "extension" and not two different types, "eip1193" and "eip6963"?  what exactly is the type "wc"-- is it wc v2 (i.e. caip25)? or is it something more general, that includes wcv1 connections (regardless of whether WC the network/company still supports the legacy form of its protocol)? I think I would be supportive of merging this as "status: draft" but I don't really think it's safe or constructive to bump that status to "final" until we've got multiple "interfaces" (native mobile, wcv2/caip25, extension) specified with sample/test dapps to show them working well with, e.g., EIP-6963, WC, MM, etc.

In other words, if people want to implement this and take it all the way, it looks like a CAIP to me, but Pedro's advice on how to do that strikes me as fairly apt!

**Comment by bumblefudge** [comment]
p.s. really excited for the authflow type "wearables"😉

File: CAIPs/caip-x.md [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+
+# dApp Implementer Steps
+Here are the detailed steps for dApps to integrate Login With Name @domain/loginwithname-wagmi-sdk connector:
+[Login With Name WAGMI SDK](https://github.com/FedericoAmura/login-with-name-wagmi-sdk)
+
+# References
+
+[EIP-4361: Sign-In with Ethereum](https://github.com/ethereum/ercs/blob/master/ERCS/erc-4361.md)
+[Login With Name WAGMI SDK](https://github.com/FedericoAmura/login-with-name-wagmi-sdk)
+
```
[the remark this answers is not in the document]
Comment (bumblefudge) [reply CAIPs/caip-x.md:388]: if you're not familar with this variant of markdown, it's how we do internal links in the CASA docs-- anywhere you put brackets around `ENSIP-5` or `JSON Schema`, it will make a link with these `hrefs`.  see the contributing.md for more details.

**Review by bumblefudge — APPROVED** [review]
I think this makes sense as a working draft to invite review and contribtion, but it won't really be done until there are multiple authflow types specified, as I said in more detail above!

**Comment by obstropolos** [comment]
Second - it meets the criteria to be approved as a draft but welcome further review as it goes through the process here.

**Review by obstropolos — APPROVED** [review]

**Comment by Sednaoui** [comment]
Great initiative. My main concern is the optional chain field when resolving a domain name. Domain names has properties to resolve to different addresses, not just a single address on a single network. 

Take ENS as an example. ENS allows a domain name to resolve to different addresses. This means I can resolve a domain name to an eth (mainnet) address, and also to an bitcoin/arbitrum/optimism address. Unfortunately, everyone (libraries, developers, etc..) defaults to resolve the name to the same address on all EVM chains, which is inaccurate for smart contract wallets, as their ownership is not defaulted to all chains. 

Having the chain field as optional may be the norm, where no one includes the chain field, leading to connectivity problems with smart contract wallets.

**Review by ustad56 — APPROVED** [review]

**Comment by FedericoAmura** [comment]
> Great initiative. My main concern is the optional chain field when resolving a domain name. Domain names has properties to resolve to different addresses, not just a single address on a single network.
> 
> Take ENS as an example. ENS allows a domain name to resolve to different addresses. This means I can resolve a domain name to an eth (mainnet) address, and also to an bitcoin/arbitrum/optimism address. Unfortunately, everyone (libraries, developers, etc..) defaults to resolve the name to the same address on all EVM chains, which is inaccurate for smart contract wallets, as their ownership is not defaulted to all chains.
> 
> Having the chain field as optional may be the norm, where no one includes the chain field, leading to connectivity problems with smart contract wallets.

Hi Sednaoui! We just opened PR #292 adding some more about how to treat the `chain` field and how it affects smart wallets

**Review by Magkoooh — APPROVED** [review]

**Review by octavio12345300 — APPROVED** [review]

**Review by octavio12345300 — APPROVED** [review]

**Comment by octavio12345300** [comment]
Aprovado
